api/utils/sshutils/ssh.go

```diff
@@ -250,3 +260,263 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
 	b := ssh.Marshal(bk)
 	return (len(a) == len(b) && subtle.ConstantTimeCompare(a, b) == 1)
 }
+
+// ConvertToPPK takes a regular RSA-formatted keypair and converts it into the PPK file format used by the PuTTY SSH client.
+// The file format is described here: https://the.earth.li/~sgtatham/putty/0.76/htmldoc/AppendixC.html#ppk
+func ConvertToPPK(priv []byte, pub []byte) ([]byte, error) {
+	// decode the private key from PEM format and extract the exponents
+	privateKeyPemBlock, rest := pem.Decode(priv)
+	if len(rest) > 0 {
+		return nil, trace.Wrap(fmt.Errorf("failed to decode private key, %v bytes left over", len(rest)))
+	}
+	privateKey, err := x509.ParsePKCS1PrivateKey(privateKeyPemBlock.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to parse private key: %T", err))
+	}
+
+	// https://the.earth.li/~sgtatham/putty/0.76/htmldoc/AppendixC.html#ppk
+	// RSA keys are stored using an algorithm-name of ‘ssh-rsa’. (Keys stored like this are also used by the updated RSA signature schemes that use
+	// hashes other than SHA-1. The public key data has already provided the key modulus and the public encoding exponent. The private data stores:
+	// mpint: the private decoding exponent of the key.
+	// mpint: one prime factor p of the key.
+	// mpint: the other prime factor q of the key. (RSA keys stored in this format are expected to have exactly two prime factors.)
+	// mpint: the multiplicative inverse of q modulo p.
+	ppkPrivateKey := new(bytes.Buffer)
+
+	// mpint: the private decoding exponent of the key.
+	// this is known as 'D'
+	mpintD, err := getRFC4251Mpint(privateKey.D)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(ppkPrivateKey, binary.BigEndian, mpintD)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("failed to write D to buffer: %T", err))
+	}
+
+	// mpint: one prime factor p of the key.
+	// this is known as 'P'
+	// the RSA standard dictates that P > Q
+	// for some reason what PuTTY names 'P' is Primes[1] to Go, and what PuTTY names 'Q' is Primes[0] to Go
+	P := privateKey.Primes[1]
```

Comment:
In this case, it might be cleaner to do this on a single line since the comment applies to both p and q.

```
p, q := privateKey.Primes[1], privateKey.Primes[0]
```